src/storage-client/src/storage_collections.rs

```diff
@@ -1824,27 +1859,167 @@ where
+        new_collection: GlobalId,
         new_desc: RelationDesc,
+        expected_version: RelationVersion,
     ) -> Result<(), StorageError<Self::Timestamp>> {
-        let mut self_collections = self.collections.lock().expect("lock poisoned");
-        let collection = self_collections
-            .get_mut(&table_id)
-            .ok_or_else(|| StorageError::IdentifierMissing(table_id))?;
-
-        // TODO(alter_table): To support changing the `RelationDesc` of sources
-        // we'll need to cancel the currently running `BackgroundCmd` that
-        // fetches recent uppers. See `BackgroundCmd::Register`.
-        if !matches!(&collection.description.data_source, DataSource::Table) {
-            return Err(StorageError::IdentifierInvalid(table_id));
-        }
+        let (data_shard, existing_write_frontier, existing_read_capabilities) = {
+            let self_collections = self.collections.lock().expect("lock poisoned");
+            let existing = self_collections
+                .get(&existing_collection)
+                .ok_or_else(|| StorageError::IdentifierMissing(existing_collection))?;
+
+            // TODO(alter_table): Support changes to sources.
+            if !matches!(&existing.description.data_source, DataSource::Table { .. }) {
+                return Err(StorageError::IdentifierInvalid(existing_collection));
+            }
 
-        collection.collection_metadata.relation_desc = new_desc.clone();
-        collection.description.desc = new_desc.clone();
+            (
+                existing.collection_metadata.data_shard,
+                existing.write_frontier.clone(),
+                existing.read_capabilities.clone(),
+            )
+        };
 
-        debug!("altered table {table_id}'s RelationDesc");
+        let persist_client = self
+            .persist
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        // Evolve the schema of this shard.
+        let diagnostics = Diagnostics {
+            shard_name: existing_collection.to_string(),
+            handle_purpose: "alter_table_desc".to_string(),
+        };
+        // We map the Adapter's RelationVersion 1:1 with SchemaId.
+        let expected_schema = expected_version.into();
+        let schema_result = persist_client
+            .compare_and_evolve_schema::<SourceData, (), T, Diff>(
+                data_shard,
+                expected_schema,
+                &new_desc,
+                &UnitSchema,
+                diagnostics,
+            )
+            .await
+            .map_err(|e| StorageError::InvalidUsage(e.to_string()))?;
+        tracing::info!(
+            ?existing_collection,
+            ?new_collection,
+            ?new_desc,
+            "evolved schema"
+        );
+
+        match schema_result {
+            CaESchema::Ok(id) => id,
+            // TODO(alter_table): Better handling of these errors.
+            CaESchema::ExpectedMismatch { .. } => {
```

Comment:
How can this happen? Also, in this case shouldn't we check if the actual schema is the one we wanted and if so continue?

Reply:
This shouldn't ever happen because the Coordinator should be the only one evolving the schema of this shard. I added a soft assertion to see if it ever does get hit, and a TODO to follow up. Right now this would just fail the `ALTER TABLE` command and the user would have to re-run it on their own.